src/dune_rules/inline_tests.ml

```diff
@@ -186,94 +187,170 @@ include Sub_system.Register_end_point (struct
+                  (Expander.expand_and_eval_set expander
+                     ~standard:(Action_builder.return []))
+              |> Action_builder.all
+            in
+            List.concat l
+          in
+          flags
+        in
         Super_context.add_alias_action sctx ~dir ~loc:(Some info.loc)
           (Alias.make ~dir runtest_alias)
-          (let exe =
-             Path.build (Path.Build.relative inline_test_dir (name ^ ext))
-           in
-           let open Action_builder.O in
-           let deps, sandbox =
-             let sandbox =
-               let project = Scope.project scope in
-               if Dune_project.dune_version project < (3, 5) then
-                 Sandbox_config.no_special_requirements
-               else Sandbox_config.needs_sandboxing
-             in
-             Dep_conf_eval.unnamed ~sandbox info.deps ~expander
-           in
-           let+ () = deps
-           and+ () = Action_builder.paths source_files
-           and+ () = Action_builder.path exe
-           and+ action =
-             match custom_runner with
-             | None ->
-               let+ flags = flags in
-               Action.run (Ok exe) flags
-             | Some runner -> (
-               let* prog =
-                 Action_builder.of_memo
-                   (Super_context.resolve_program ~dir sctx ~loc:(Some loc)
-                      runner)
-               and* flags = flags in
-               let action =
-                 Action.run prog (Path.reach exe ~from:(Path.build dir) :: flags)
-               in
-               (* jeremiedimino: it feels like this pattern should be pushed
-                  into [resolve_program] directly *)
-               match prog with
-               | Error _ -> Action_builder.return action
-               | Ok p -> Action_builder.path p >>> Action_builder.return action)
-           in
-           let run_tests = Action.chdir (Path.build dir) action in
-           Action.Full.make ~sandbox
-           @@ Action.progn
-                (run_tests
-                :: List.map source_files ~f:(fun fn ->
-                       Action.diff ~optional:true fn
-                         (Path.Build.extend_basename
-                            (Path.as_in_build_dir_exn fn)
-                            ~suffix:".corrected")))))
+          (let open Action_builder.O in
+          let actions =
+            let* partitions_flags =
+              match partitions_flags with
+              | None -> Action_builder.return [ None ]
+              | Some _ ->
+                let+ partitions =
+                  Action_builder.lines_of (Path.build partition_file)
+                in
+                List.map ~f:(fun x -> Some x) partitions
+            in
+            let+ action =
+              List.map partitions_flags ~f:(fun p -> action mode (flags p))
+              |> Action_builder.all
+            in
+            action
+          in
+          let+ action = actions
```

Comment:
You don't need to use conc on these actions. You can just add them individually to the alias for the same effect. It's only within a single partition that you need to use conc to call diff on all the sources inside that partition.

Reply:
There is currently not relationship between partitions and source files. We need to run all partition before we can diff sources

Reply:
I see. Well you can still run them in parallel without the use of conc. I would suggest to do that as a first step and leave the conc stuff for a subsequent PR.

Reply:
I'm not sure what your asking for here. Could you be more specific ? What actions are you talking about exactly ? Diffs need to run after the all the tests.

Reply:
Correct me if I'm wrong, but this is what you're doing currently:

1. Execute the runner to the get the list of partitions
2. Generate a list of actions from this list of partitions
3. Run these actions concurrently.
4. Then diff them all concurrently

Now here's what I suggested:

1. Execute the runner to get the list of partitions
2. For every partition, create an action that will execute the partition. You can attach all these actions to some alias `partitions` in the inline tests directory
3. Create another action that will depend on the `partitions` alias and will then run the diff (concurrently, once Ali's PR is merged). Add this action to the `runtest` alias.

The advantage of my proposal is that passing partitions will not be re-executed.

But now that I think about it, you might need to go through contortions to make sure that executing the test to get the list of partitions will be done when loading the `.inline-tests` sub directory rather than the directory where you're building the inline tests themselves. I'm not so sure it's worth it anymore.

Reply:
> The advantage of my proposal is that passing partitions will not be re-executed.

I'm not sure what you mean here.
Partitions will have have to be recomputed every time the executable is regenerated.
Maybe you mean that actions responsible for running each partition would not have to be recomputed ?

I think what you're proposing is exactly what I commented about in https://github.com/ocaml/dune/issues/1516#issuecomment-1379538124
> I've tried to implement the feature and got stuck realizing that I want to declare an arbitrary number of alias action (add_alias_action) based on a list of partition previously computed. I think this cannot currently be done, is it correct ?

And I thought it was impossible, but maybe it's possible with contortions